solution checksum and serialization service.

heejaechang opened the pull request:
actual commit with changes.
https://github.com/dotnet/roslyn/pull/12771/commits/d0b58a20bc1586ff68bb49a1c1f23e56169662b9

work left for this area.
https://github.com/dotnet/roslyn/issues/12770

heejaechang commented:
@mattwar @CyrusNajmabadi @dotnet/roslyn-analysis @dotnet/roslyn-ide @mavasani @srivatsn can you take a look? thank you.

heejaechang commented:
@panopticoncentral adding paul.

heejaechang commented:
test will fail since PR doesnt contain csproj/vbproj file changes for added files

panopticoncentral commented on src/Workspaces/Core/Portable/Execution/Serialization/SnapshotStorage.cs:
```diff
@@ -0,0 +1,461 @@
+    /// <summary>
+    /// This is collection of hierarchical checksum trees
+    /// </summary>
+    internal class SnapshotStorages
```
>SnapshotStorages [](start = 19, length = 16)

Should be SnapshotStorage?

heejaechang replied:
renamed to ChecksumTreeCollections

panopticoncentral commented on src/Workspaces/Core/Portable/Execution/Extensions.cs:
```diff
@@ -0,0 +1,107 @@
+
+namespace Microsoft.CodeAnalysis.Execution
+{
+    // REVIEW: should there be auto retry mechanism such as if out of proc run fails, it automatically calls inProc one?
```
>// REVIEW: should there be auto retry mechanism such as if out of proc run fails, it automatically calls inProc one? [](start = 4, length = 116)

This comment doesn't seem to be connected to anything in this file?

heejaechang replied:
yep. deleted

panopticoncentral commented on src/Workspaces/Core/Portable/Execution/Serialization/IReferenceSerializationService.cs:
```diff
@@ -0,0 +1,35 @@
+    /// This lets consumer to get to inner temporary storage that references use
+    /// as its shadow copy storage
+    /// </summary>
+    internal interface ISupportTemporaryStorage
```
>ISupportTemporaryStorage [](start = 23, length = 24)

Shouldn't this be in its own file?

heejaechang replied:
done

panopticoncentral commented on src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs:
```diff
@@ -0,0 +1,549 @@
+            WriteMvidTo((ModuleMetadata)metadata, writer, cancellationToken);
+        }
+
+        private void WriteMvidTo(ModuleMetadata metadata, ObjectWriter writer, CancellationToken cancellationToken)
```
>CancellationToken cancellationToken [](start = 79, length = 35)

Not used.

heejaechang replied:
added cancellation check

panopticoncentral commented on src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs:
```diff
@@ -0,0 +1,549 @@
+            }
+        }
+
+        private void WriteMvidsTo(PortableExecutableReference reference, ObjectWriter writer, CancellationToken cancellationToken)
```
>CancellationToken cancellationToken [](start = 94, length = 35)

Not used once you remove the param from WriteMvidTo

heejaechang replied:
made inner one to check cancellation

panopticoncentral commented on src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs:
```diff
@@ -0,0 +1,549 @@
+            throw ExceptionUtilities.UnexpectedValue(kind);
+        }
+
+        private void WriteTo(MetadataReferenceProperties properties, ObjectWriter writer, CancellationToken cancellationToken)
```
>CancellationToken cancellationToken [](start = 90, length = 35)

Not used.

heejaechang replied:
added cancellation check

panopticoncentral commented on src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs:
```diff
@@ -0,0 +1,549 @@
+            writer.WriteBoolean(properties.EmbedInteropTypes);
+        }
+
+        private MetadataReferenceProperties ReadMetadataReferencePropertiesFrom(ObjectReader reader, CancellationToken cancellationToken)
```
>CancellationToken cancellationToken [](start = 101, length = 35)

Not used.

heejaechang replied:
added cancellation check.

panopticoncentral commented on src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs:
```diff
@@ -0,0 +1,549 @@
+            stream.Write(content, 0, content.Length);
+        }
+
+        private void WriteTo(ModuleMetadata metadata, ObjectWriter writer, CancellationToken cancellationToken)
```
>CancellationToken cancellationToken [](start = 75, length = 35)

Not used

heejaechang replied:
same here, cancellation check is added to callee

panopticoncentral commented on src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs:
```diff
@@ -0,0 +1,549 @@
+            WriteTo(metadataReader, writer, cancellationToken);
+        }
+
+        private void WriteTo(MetadataReader reader, ObjectWriter writer, CancellationToken cancellationToken)
```
>CancellationToken cancellationToken [](start = 73, length = 35)

Not used.

heejaechang replied:
same

panopticoncentral commented on src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs:
```diff
@@ -0,0 +1,549 @@
+            return new MetadataReferenceProperties(kind, aliases, embedInteropTypes);
+        }
+
+        private void WriteTo(Metadata metadata, ObjectWriter writer, CancellationToken cancellationToken)
```
>CancellationToken cancellationToken [](start = 69, length = 35)

Not used.

heejaechang replied:
same

panopticoncentral commented on src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs:
```diff
@@ -0,0 +1,549 @@
+            // TODO: what I should do with documentation provider? it is not exposed outside
+        }
+
+        private void WritePortableExecutableReferenceMetadataTo(PortableExecutableReference reference, ObjectWriter writer, CancellationToken cancellationToken)
```
>CancellationToken cancellationToken [](start = 124, length = 35)

Not used.

heejaechang replied:
now WriteTo check cancellation

panopticoncentral commented on src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs:
```diff
@@ -0,0 +1,549 @@
+        private sealed class PinnedObject : IDisposable
+        {
+            private readonly GCHandle _gcHandle;
+            private readonly long _length;
```
> _length [](start = 33, length = 8)

This field is never used.

panopticoncentral commented on src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs:
```diff
@@ -0,0 +1,549 @@
+                return _gcHandle.AddrOfPinnedObject();
+            }
+
+            private void Dispose(bool disposing)
```
>disposing [](start = 38, length = 9)

Parameter isn't used.

panopticoncentral commented on src/Workspaces/Core/Portable/Execution/Serialization/Asset.cs:
```diff
@@ -0,0 +1,148 @@
+        }
+    }
+
+    internal sealed class MetadataReferenceAsset : Asset
```
>MetadataReferenceAsset [](start = 26, length = 22)

Aren't classes like these split out into their own files?

heejaechang replied:
I have work item to clean up some assets. I will do that when I do that work.

panopticoncentral commented:
:+1: But there are specific questions in there that someone with more Roslyn experience than I have should answer or at least review...

heejaechang commented:
@panopticoncentral I addressed your feedback in this commit https://github.com/dotnet/roslyn/pull/12771/commits/2bc40863d782a9b44ef2e473ae29430a539272b6

panopticoncentral commented:
:+1:

heejaechang commented:
this PR is merged to this
https://github.com/dotnet/roslyn/pull/12774
